### ISSUE TITLE
Notifications are not delayed by times.begin after the expiry of a downtime.

### ISSUE THREAD

**Issue by TQQEU** [Icinga/icinga2#9526]
## Describe the bug

We have the following problem in the environment we service:

We have several types of notifications, including emails and a ticket system.
We would like to include a link to the issued ticket in the eMails sent, which is why we use the following mechanism to delay the sending of the mail:

``` icinga2
apply Notification "N-Mail" to Service {
    #[...]

    if (vars.notification_ticket)  {
        times.begin = 30s
    }

    #[...]
}
```

This mechanism allows us in almost all cases to delay the sending of mails until 30 seconds after the ticket has been created, so that we can make use of a comment that is created when the ticket is created (usually within 3 seconds).

That does not work after Downtimes.

In out case that means we send out an olt ticket-number in the Notification-Mails.

## To Reproduce

1. Setup Icinga2 to send Notifications using the mechanism mentioned above.
2. Set a Downtime for a Service Object. That Service is "OK".
3. Whithin that Downtime the Service becomes Hard-Critical.
4. After the downtime has ended, the suppressed notifications will be sent out immedeately, ignoring the setting `times.begin`

## Expected behavior

times.begin also delays Notifications after a downtime

## Your Environment

Include as many relevant details about the environment you experienced the problem in

* Version used (`icinga2 --version`): `r2.13.3-1`
* Operating System and version: `Red Hat Enterprise Linux release 8.6 (Ootpa)`
* Enabled features (`icinga2 feature list`): `api checker ido-mysql influxdb mainlog notification`

**Comment by bobapple** [comment]
ref/IP/43340

**Comment by Al2Klimov** [comment]
# Task

Reproduce. (For now just reproduce! Maybe you'll have more success than I.)

**Comment by julianbrost** [comment]
`times.begin` is checked relative to the problem begin, so you should see that behavior if the 30s pass within the downtime: https://github.com/Icinga/icinga2/blob/a958a735d7180d68829cdc0bb18f43d9e597c980/lib/icinga/notification.cpp#L286

**Comment by Al2Klimov** [comment]
Hello @TQQEU!

Did I understand correctly? You open a ticket via notification and 30s later you send a notification eMail with the ticket nr?

What about sleeping 30s first in your eMail notification?

Best,
A/K

**Comment by TQQEU** [comment]
> What about sleeping 30s first in your eMail notification?

Hello,
we've thought about it.
People here are afraid that - in case of an event where many mails have to go out - too many processes will remain open (waiting). There is a little difference in sending a Mail out in <1s and sleeping 30s before it.

**Comment by Al2Klimov** [comment]
I assure you: a sleep process and its parent shell won't consume too much resources.

**Comment by Al2Klimov** [comment]
@julianbrost Aren't our docs clear enough on the status quo?

https://github.com/Icinga/icinga2/blob/ae32b3cbbdcaabd71a214eff726f85bc598f36ee/doc/03-monitoring-basics.md#notification-delay-

> when the notification is due (**the object reaching the HARD state**), but after a certain period

**Comment by julianbrost** [comment]
If you know the behavior, you can recognize it in the docs, but can you derive the behavior from the docs without doubt? Let me give you another example: Imagine a notification with `times.begin = 1h` and `times.end = 2h` and a hard critical state happening 3h before a downtime ends. Is there anything in the documentation that makes it clear what would happen (no notification)?

The statement near the end of the [previous section](https://github.com/Icinga/icinga2/blob/ae32b3cbbdcaabd71a214eff726f85bc598f36ee/doc/03-monitoring-basics.md#notification-escalations-) is also only correct as long as there's no downtime:
> If the problem does not get resolved nor acknowledged preventing further notifications, the escalation-sms-1st-level user will be escalated 1h after the initial problem was notified, but only for one hour (2h as end key for the times dictionary).

If the corresponding hard state occurs during a downtime, it may be less than an hour apart.

So the docs should explicitly mention the interaction with downtimes.

**Comment by TQQEU** [comment]
When we discussed the problem further internally we came to the following points that might be of interest:

In general we have an environment with about 40.000 Services and about 2.000 Hosts.

The problems we have observed always occur in a situation of higher load, since the checks for which this behaviour is relevant usually only occur in the case of a major fault.

Of course, we welcome it if the current behaviour is made clear in the documentation.

As an alternative solution, we would appreciate it if comments added via the API (in our case by a ticket number) were immediately available for the next API-query (in our case for assembling the eMail), even under heavy load. Then we could save ourselves the additional 30 second wait.
